### PR TITLE
Add new PnP Samples for the DigitalTwinClient

### PR DESCRIPTION
**Pull request by barustum**
## Purpose

* Add new PnP Samples for the DigitalTwinClient
* Move PnPServiceSamples from the azure-iot-sdk-csharp (preview branch) to this repository, as all samples should live in one place
* Move PnPDeviceSamples from the azure-iot-sdk-csharp (preview branch) to this repository, as all samples should live in one place

## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[x] Other... Please describe: Add Samples
```

### CODE REVIEW

File: iot-hub/Samples/device/PnpDeviceSamples/TemperatureController/TemperatureControllerSample.cs [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+
+            await UpdateDeviceInformationAsync(cancellationToken);
+            await SendDeviceMemoryAsync(cancellationToken);
+            await SendDeviceSerialNumberAsync(cancellationToken);
+
+            bool temperatureReset = true;
+            _maxTemp[Thermostat1] = 0d;
+            _maxTemp[Thermostat2] = 0d;
+
+            await Task.Run(async () =>
```
Comment (azabbasi) [inline iot-hub/Samples/device/PnpDeviceSamples/TemperatureController/TemperatureControllerSample.cs:93]: The context of this method is async Task, I don't think you need Task.Run here do you?

File: iot-hub/Samples/device/PnpDeviceSamples/TemperatureController/TemperatureControllerSample.cs [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+                            $" maxTemp={report.maxTemp}, minTemp={report.minTemp}, avgTemp={report.avgTemp}, startTime={report.startTime.LocalDateTime}, " +
+                            $"endTime={report.endTime.LocalDateTime}");
+
+                        byte[] responsePayload = Encoding.UTF8.GetBytes(JsonConvert.SerializeObject(report));
+                        return await Task.FromResult(new MethodResponse(responsePayload, (int)StatusCode.Completed));
+                    }
+
+                    _logger.LogDebug($"Command: component=\"{componentName}\", no relevant readings found since {sinceInDateTimeOffset.LocalDateTime}, " +
+                        $"cannot generate any report.");
+                    return await Task.FromResult(new MethodResponse((int)StatusCode.NotFound));
```
Comment (azabbasi) [inline iot-hub/Samples/device/PnpDeviceSamples/TemperatureController/TemperatureControllerSample.cs:178]: I think here you can just do `return new MethodResponse((int)StatusCode.NotFound)`

File: iot-hub/Samples/device/PnpDeviceSamples/TemperatureController/TemperatureControllerSample.cs [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+                    callbackNotInvoked = false;
+                }
+            }
+
+            if (callbackNotInvoked)
+            {
+                _logger.LogDebug($"Property: Received a property update that is not implemented by any associated component.");
+            }
+
+            await Task.CompletedTask;
```
Comment (azabbasi) [inline iot-hub/Samples/device/PnpDeviceSamples/TemperatureController/TemperatureControllerSample.cs:210]: I don't think you need this at line 210.

File: iot-hub/Samples/device/PnpDeviceSamples/TemperatureController/TemperatureControllerSample.cs [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+        // The desired property update callback, which receives the target temperature as a desired property update,
+        // and updates the current temperature value over telemetry and property update.
+        private async Task TargetTemperatureUpdateCallbackAsync(TwinCollection desiredProperties, object userContext)
+        {
+            const string propertyName = "targetTemperature";
+            string componentName = (string)userContext;
+
+            bool targetTempUpdateReceived = PnpHelper.TryGetPropertyFromTwin(desiredProperties,
+                                                                             propertyName,
+                                                                             out double targetTemperature,
```
Comment (azabbasi) [inline iot-hub/Samples/device/PnpDeviceSamples/TemperatureController/TemperatureControllerSample.cs:222]: if you are breaking the lines, I suggest you bring the first parameter down to a new line and use only 1 tab.
```
bool targetTempUpdateReceived = PnpHelper.TryGetPropertyFromTwin(
        desiredProperties,
        propertyName,
        out double targetTemperature,
        componentName);
```

File: iot-hub/Samples/device/PnpDeviceSamples/Thermostat/Parameter.cs [hunk 5]
```diff
@@ -0,0 +1,80 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
```
Comment (azabbasi) [inline iot-hub/Samples/device/PnpDeviceSamples/Thermostat/Parameter.cs:2]: This class is identical to that of the TemperatureController ... can we reuse that one?

File: iot-hub/Samples/device/PnpDeviceSamples/Thermostat/ThermostatSample.cs [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+                        startTime = filteredReadings.Keys.Min(),
+                        endTime = filteredReadings.Keys.Max(),
+                    };
+
+                    _logger.LogDebug($"Command: MaxMinReport since {sinceInDateTimeOffset.LocalDateTime}:" +
+                        $" maxTemp={report.maxTemp}, minTemp={report.minTemp}, avgTemp={report.avgTemp}, " +
+                        $"startTime={report.startTime.LocalDateTime}, endTime={report.endTime.LocalDateTime}");
+
+                    byte[] responsePayload = Encoding.UTF8.GetBytes(JsonConvert.SerializeObject(report));
+                    return await Task.FromResult(new MethodResponse(responsePayload, (int)StatusCode.Completed));
```
Comment (azabbasi) [inline iot-hub/Samples/device/PnpDeviceSamples/Thermostat/ThermostatSample.cs:150]: same here, no need to do Task.FromResult

File: iot-hub/Samples/device/PnpDeviceSamples/Thermostat/Parameter.cs (hunk 5 above)
Comment (barustum) [reply iot-hub/Samples/device/PnpDeviceSamples/Thermostat/Parameter.cs:2]: We are trying to keep all samples independent of each other, so a user can choose to run any sample without the context of the solution.

**Pull request merged**